spec/controllers/miq_request_controller_spec.rb

```diff
@@ -266,6 +266,52 @@
     end
   end
 
+  context 'showing details of a retirement task' do
+    before do
+      stub_admin
+      EvmSpecHelper.create_guid_miq_server_zone
+    end
+
+    let(:vm1)     { FactoryGirl.create(:vm_cloud) }
+    let(:vm2)     { FactoryGirl.create(:vm_cloud) } # not part of the stack
+    let(:stack)   { FactoryGirl.create(:orchestration_stack).tap { |stack| stack.direct_vms = [vm1] } }
+    let(:service) { FactoryGirl.create(:service_orchestration).tap { |service| service.add_resource!(stack) } }
+    let(:request) { FactoryGirl.create(:miq_service_retirement_request, :options => {:src_ids => [service.id]}) }
+
+    let(:angular_payload) { { :model_name => 'Vm', :parent_id => service.id.to_s } }
+    let(:angular_additional) do
+      {
+        :model             => 'Vm',
+        :parent_id         => service.id.to_s,
+        :parent_class_name => 'ServiceOrchestration',
+        :view_suffix       => 'OrchestrationStackRetireRequest',
+        :display           => 'main',
+        :gtl_type          => 'list'
+      }
+    end
+
+    it 'angular for grid with affected VMs is properly initialized' do
+      payload = angular_payload.update(:report_data_additional_options => angular_additional)
+      expect(controller).to receive(:prov_set_show_vars).once.and_call_original
+
+      # Verify Rails correctly initializes Angular which will then perform POST /report_data to fetch the VMs.
+      expect_any_instance_of(GtlHelper).to receive(:render_gtl).with match_gtl_options(**payload)
+
+      get :show, :params => {:id => request.id}
+      expect(response.status).to eq(200)
+    end
+
+    it 'angular for grid with affected VMs gets the correct VMs with XHR call' do
+      payload = angular_payload.update(:additional_options => angular_additional)
+      post :report_data, :params => payload
+      expect(response.status).to eq(200)
+
+      # Verify Angular got correct VMs when sending the payload as set by previous test.
+      obtained = JSON.parse(response.body)
```

Comment:
YAGNI -> `response.parsed_body`

Reply:
Dunno, `.parsed_body` is string for me no matter what headers I set. Even
<pre>
post :report_data, :params => payload, <b>:format => :json</b>
</pre>

won't help.. I see other people also need workaround to get this working https://github.com/rails-api/active_model_serializers/issues/1437#issuecomment-172415310